### PR TITLE
Real interferometer

### PR DESCRIPTION
**Pull request by ziofil**
**Context:**
It would be nice to have a trainable *real* interferometer (i.e. one that doesn't mix q and p)

**Description of the Change:**
I added a new gate: `RealInterferometer`

**Benefits:**
We have it now :)

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: tests/test_utils/test_opt.py [hunk 1]
```diff
@@ -214,6 +247,39 @@ def cost_fn():
     assert np.allclose(-cost_fn(), 0.0625, atol=1e-5)
 
 
+def test_learning_four_mode_RealInterferometer():
+    """Finding the optimal Interferometer to make a NOON state with N=2"""
+    np.random.seed(11)
```
Comment (sduquemesa) [inline tests/test_utils/test_opt.py:252]: Same comment in regards to making tests deterministic

File: mrmustard/math/math_interface.py [hunk 2]
```diff
@@ -933,13 +872,20 @@ def random_symplectic(self, num_modes: int = 1, max_r: float = 1.0) -> Tensor:
...
-        """A random orthogonal matrix in :math:`O(2*num_modes)`."""
-        if num_modes == 1:
+    @staticmethod
+    def random_orthogonal(N: int) -> Tensor:
+        """A random orthogonal matrix in :math:`O(N)`."""
+        if N == 1:
+            return np.array([[1.0]])
+        return ortho_group.rvs(dim=N)
+
+    def random_unitary(self, N: int) -> Tensor:
```
Comment (sduquemesa) [inline mrmustard/math/math_interface.py:882]: I would suggest a (very minor) refactoring here just for the sake of readability:
```suggestion
    def random_unitary(self, N: int) -> Tensor:
        """a random unitary matrix in :math:`U(N)`"""
        if N == 1:
            return self.exp(1j * np.random.uniform(size=(1, 1)))
        return unitary_group.rvs(dim=N)
```

**Review by sduquemesa — COMMENTED** [review]
Hi @ziofil, this is looking good from my side! ✨ I left some small suggestions and questions on the implementation.

On the testing side, I found a pattern related to making tests deterministic that was common  in SF and perhaps we could try to avoid bringing into Mr Mustard. 🔍

File: tests/test_utils/test_opt.py [hunk 3]
```diff
@@ -214,6 +247,39 @@ def cost_fn():
...
+                    * np.array([[0, 0, 1 / np.sqrt(2)], [0, 0, 0], [1 / np.sqrt(2), 0, 0]])
+                )
+            )
+            ** 2
+        )
+
+    opt = Optimizer(symplectic_lr=0.5, euclidean_lr=0.01)
+
+    opt.minimize(cost_fn, by_optimizing=[circ], max_steps=1000)
+    assert np.allclose(-cost_fn(), 0.0625, atol=1e-5)
```
[the remark this answers is not in the document]
Comment (ziofil) [reply tests/test_utils/test_opt.py:280]: it's 1/16

File: tests/test_utils/test_opt.py [hunk 4]
```diff
@@ -181,6 +189,31 @@ def cost_fn():
...
+    state_in = Vacuum(num_modes=2)
+
+    def cost_fn():
+        amps = (state_in >> circ).ket(cutoffs=[2, 2])
+        return -tf.abs(amps[1, 1]) ** 2 + tf.abs(amps[0, 1]) ** 2
+
+    opt = Optimizer(orthogonal_lr=0.5, euclidean_lr=0.01)
+
+    opt.minimize(cost_fn, by_optimizing=[circ], max_steps=1000)
+    assert np.allclose(-cost_fn(), 0.25, atol=1e-5)
```
[the remark this answers is not in the document]
Comment (ziofil) [reply tests/test_utils/test_opt.py:214]: yes!

File: tests/test_utils/test_opt.py [hunk 5]
```diff
@@ -181,6 +189,31 @@ def cost_fn():
     assert np.allclose(-cost_fn(), 0.25, atol=1e-5)
 
 
+def test_learning_two_mode_RealInterferometer():
+    """Finding the optimal Interferometer to make a pair of single photons"""
+    np.random.seed(11)
```
[the remark this answers is not in the document]
Comment (ziofil) [reply tests/test_utils/test_opt.py:194]: I agree with you. Maybe this is not the right PR to start working on these issues though.

**Review by sduquemesa — APPROVED** [review]

**Pull request merged**